[mlir-tblgen] Relax builder ambiguity check

RoboTux opened the pull request:
The mlir-tblgen tool prevents the parameter of the build() constructor
for the first default-valued attribute of an operation from having a
default value to avoid ambiguity with the corresponding build()
constructor taking unwrapped value. However it does so even when earlier
wrapped unwrappable attribute would lift the ambiguity. This commit
relax the logic accordingly, which allows to remove a manual constructor
in Arith dialect.

RoboTux committed:
[mlir-tblgen] Relax builder ambiguity check

The mlir-tblgen tool prevents the parameter of the build() constructor
for the first default-valued attribute of an operation from having a
default value to avoid ambiguity with the corresponding build()
constructor taking unwrapped value. However it does so even when earlier
wrapped unwrappable attribute would lift the ambiguity. This commit
relax the logic accordingly, which allows to remove a manual constructor
in Arith dialect.

kuhar commented on mlir/tools/mlir-tblgen/OpDefinitionsGen.cpp:
```diff
@@ -3088,6 +3088,26 @@ void OpEmitter::buildParamList(SmallVectorImpl<MethodParameter> &paramList,
+  // Check if parameters besides default valued one are enough to distinguish
+  // between builders with wrapped and unwrapped arguments.
+  bool hasBuilderAmbiguity = true;
+  for (int i = 0; i < op.getNumArgs(); ++i) {
```
nit: don't recompute the end after every iteration: https://llvm.org/docs/CodingStandards.html#don-t-evaluate-end-every-time-through-a-loop

You could probably also turn this into a range for loop

krzysz00 commented:
Seems reasonable conceptually aside the nits above

RoboTux committed:
Use range for loop

RoboTux committed:
Merge remote-tracking branch 'upstream/main' into relax_builder_overlap_tablegen_check

RoboTux committed:
Remove no longer needed Tosa Rescale constructor

A constructor with default value for both input_unsigned and
output_unsigned is now being generated with both wrapped and unwrapped
parameters.

jpienaar reviewed (approved):
This looks good, thanks. I'll do a quick check to see if I see any other failures on larger test set later today.

jpienaar commented:
I see some failures MHLO side ( out-of-line definition of 'build' does not match any declaration ), so seems like this is not just avoiding ambiguity but may also skipping generating some build methods that was generated before.

jpienaar commented:
Yes, so this results in generating different number of build methods than before in the header, but not the bodies. You can check in build of

tensorflow/compiler/xla/mlir_hlo/_objs/mlir_hlo/hlo_ops

jpienaar reviewed (changes requested):
Ensure generating same number of build methods in header as body.

jpienaar reviewed (approved):
OK, I had time to look more. The confusing thing was the user provided builder was not preferred but that's not new here. Post that only 1 change I noticed, seems a different build method being called which resulted in `DefaultValuedOptionalAttr` not being set. But that doesn't seem to be a difference introduce here.
